### PR TITLE
fix: use async config loading if available

### LINKED ISSUE #824: ESM configs can't load
## **BEFORE YOU SUBMIT** please read the following:

**I'm submitting a bug report**

**Node Version:**
13.8

**Webpack Version:**
4.41

**Babel Core Version**:
7.8

**Babel Loader Version**:
8.0.6

**Please tell us about your environment:**
Linux

**Current behavior:**

In case of using `type=module` in `package.json` or `.mjs` extension for ESM configuration file Babel throws an error:

```text
babel.config.js: Error while loading config - You appear to be using a native ECMAScript module configuration file, which is only supported when running Babel asynchronously.
```

**Expected/desired behavior:**

ESM configs are supporter by Babel and should load/work.

This [line in index.js](https://github.com/babel/babel-loader/blob/master/src/index.js#L161) uses synchronous `loadPartialConfig` which should be `loadPartialConfigAsync` instead and `await` (parent function is async already).

### PR DESCRIPTION
**Pull request by the-spyke**
**Please Read the [CONTRIBUTING Guidelines](https://github.com/babel/babel-loader/blob/master/CONTRIBUTING.md)**
*In particular the portion on Commit Message Formatting*

**Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

**What is the current behavior?** (You can also link to an open issue here)

Fixes #824 

**What is the new behavior?**

If `@babel/core` is `>=7.8.0`, then ESM config will be found and loaded asynchronously.

**Does this PR introduce a breaking change?**
- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the following...

* Impact:
* Migration path for existing applications:
* Github Issue(s) this is regarding:

**Other information**:

### CODE REVIEW

**Comment by the-spyke** [comment]
I have no idea why [sourcemaps test fails](https://ci.appveyor.com/project/babel/babel-loader/build/job/68un2k0x9btg6nb0#L79) on Windows in my branch.

File: src/index.js [hunk 1]
```diff
@@ -158,7 +158,8 @@ async function loader(source, inputSourceMap, overrides) {
     );
   }
 
-  const config = babel.loadPartialConfig(injectCaller(programmaticOptions));
+  const config = await (babel.loadPartialConfigAsync ||
```
Comment (JLHwung) [inline src/index.js]: nit: Can you add a comment here? something like
```js
// babel.loadPartialConfigAsync is available in v7.8.0+
```

Comment (the-spyke) [reply src/index.js]: Done

**Comment by JLHwung** [comment]
the AppVeyor CI error is not directly related to this change, but babel changed some path related behaviour from 7.2.0 to 7.8.0

**Comment by the-spyke** [comment]
@JLHwung `C:/projects/babel-loader/test/fixtures/basic.js` definitely looks wrong.

File: test/loader.test.js [hunk 2]
```diff
@@ -127,3 +127,38 @@ test.cb(
     });
   },
 );
+
+test.cb("should load ESM config files", t => {
```
Comment (nicolo-ribaudo) [inline test/loader.test.js]: Can we differentiate this test based on `process.version`?

Comment (the-spyke) [reply test/loader.test.js]: I was thinking about it, but I don't know if Node team will backport ESM to 12 or even 10 in the future and this test begin to fail. But okay, I'll redo it with semver check of Node version. As I remember `13.3` was the first one without `--experimental` flag.

Comment (nicolo-ribaudo) [reply test/loader.test.js]: Let's only test in on 13.3 for now. If they'll backport it we can update the test.

Comment (the-spyke) [reply test/loader.test.js]: Done

**Comment by the-spyke** [comment]
So, I've tried to overcome ESM loading on Window. But even in case of using [file urls according to Node](https://github.com/nodejs/node/issues/31710) it fails. This time because of `request@1.12.0` even on Linux:

```text
Module build failed (from ./lib/index.js):
   Error: Cannot find module 'file:///home/user/babel-loader/test/fixtures/babelrc.mjs' from '/home/user/babel-loader'
       at /home/user/babel-loader/node_modules/resolve/lib/async.js:97:35
       at processDirs (/home/user/babel-loader/node_modules/resolve/lib/async.js:244:39)
       at isdir (/home/user/babel-loader/node_modules/resolve/lib/async.js:251:32)
       at /home/user/babel-loader/node_modules/resolve/lib/async.js:23:69
       at FSReqCallback.oncomplete (fs.js:166:21)
```

**Comment by nicolo-ribaudo** [comment]
The windows bug should be fixed by https://github.com/babel/babel/pull/11193

**Comment by the-spyke** [comment]
@nicolo-ribaudo So, what do we do? Rollback to `pathToUrl` and wait till it's fixed, or rollback and merge now?

**Comment by nicolo-ribaudo** [comment]
Let's wait for now, I'll release a patch version when there Babel PR is fixed.

**Comment by nishantjr** [comment]
Can this be merged now?

**Comment by nicolo-ribaudo** [comment]
Thanks!

**Pull request merged**